### DISCUSSION TITLE
0.19.0 became very slow in development environment

### DISCUSSION THREAD

**Discussion by oeyoews** [wxt-dev/wxt#866] [General]
I upgraded from 0.18.15 to 0.19.0, and I obviously felt that the compilation speed in the development environment became very slow. Does anyone have similar experience?

**Comment by aklinker1** [comment]
Does using `entrypointLoader: 'jiti'` fix the performance issues?

**Reply by oeyoews** [reply]
> Does using `entrypointLoader: 'jiti'` fix the performance issues?

I haven't used this yet, I'll try it tomorrow

**Reply by aklinker1** [reply]
[Some others have reported errors with 0.19.0's `vite-node` loader](https://github.com/wxt-dev/wxt/issues/864), and I've been releasing some alpha versions for them to test.

I just released `wxt@0.19.1-alpha3` that includes a performance change to tree-shake out any code that isn't related to the entrypoint options. It should be as fast or faster than v0.18 and below.

_Edit: `v0.19.1` is officially out, give that a go instead._

```sh
pnpm i wxt@0.19.1
```

**Comment by jaek-s** [comment]
Hey! I'm jumping in here because I'm experiencing the same issue. For my extension, v0.18 will give me the "Opened browser" message in under 2s or so. After upgrading wxt to v0.19.4 will take longer than 6s, usually closer to 8s before the "Opened browser" message appears.

I tried adding `entrypointLoader: 'jiti'`, but that didn't seem to have any effect.

I'm really not sure what could be causing the issue. Maybe some dependency I have, or just the size or complexity of my project? I noticed that there was some slowdown after upgrading some of the [examples](https://github.com/wxt-dev/examples) to V0.19 and downgrading [github-better-line-counts](https://github.com/aklinker1/github-better-line-counts) to v0.18 resulted in some speed up. Nothing as dramatic as what I'm seeing, to be sure.

Let me know if there's anything I can do to help track this down!

**Reply by 1natsu172** [reply]
I am wondering if the build is slow, or if the browser startup process is slow, or both.

FYI: I tried too on example and the score is slightly but not that bad on my machine.

> ← WXT@0.19.4 | WXT@0.18.5 →
![image](https://github.com/user-attachments/assets/190a2f74-fd83-4123-9752-7a7a074b7c5d)

**Reply by aklinker1** [reply]
I've also noticed that browser startup is taking longer... Only thing I changed is I upgraded some dependencies. I can run a bisect to try and find what commit caused this

**Reply by aklinker1** [reply]
Bisect wasn't really feasible (too much variation to detect the problematic commit), so I did a small analysis of how long it takes different versions of WXT to open the browser.

<img width="602" alt="Screenshot 2024-08-12 at 2 31 04 PM" src="https://github.com/user-attachments/assets/47ecb707-0fc7-4cdc-93cd-79366e2bc225">

>  I couldn't figure out how to add stdev based error bars in google sheets lol, so excuse my questionable charting skills

https://docs.google.com/spreadsheets/d/1Z87paldvcokI5tVtJpQ1yK3ughv8SPtT3IOfPbYLrgM/edit?usp=sharing

So it looks like v0.18.7 and v0.19.0 both introduced some form of slow-down. I'll do another benchmark looking at commits for those versions, and post another update here.

**Reply by aklinker1** [reply]
Alright, there we go. Got a couple of bad commits.
<img width="1409" alt="Screenshot 2024-08-12 at 6 27 30 PM" src="https://github.com/user-attachments/assets/74ea45bb-30cf-4f11-9980-3002a27a49fa">
- 48783ba
- 5eb5be7 (or f1bb7cb?)
- And maybe a0971ca?

**Reply by 1natsu172** [reply]
Hmm...It is weird that there are no diffs in `web-ext.ts` since v0.18.6 that might have a performance impact.

I looked at the performance profiling on v0.19.5 with `node --inspect-brk ./node_modules/wxt/bin/wxt.mjs` in `wxt/packages/wxt-demo`, it seems like there are so many stacks of `resolveConfig` and `resolveWxtUserModules`.

(v0.18.7 & v19.0.0 would have included changes related to `@wxt-dev/module-<>` and auto-icon, right?)

![image](https://github.com/user-attachments/assets/bb89974e-8a19-4dce-b468-a0cfa1cd5751)

Just another viewpoint, hypothetical…isn't this simply performance degradation due to the accumulated processes? If there is no problem with dependency modules, we may suspect cpu thread depletion or memory pressure.

**Reply by aklinker1** [reply]
Could be! That would explain why it goes from almost no variation in duration to varying wildly. You can see the minimum duration doesn't really change between versions - so it can be as fast as it was originally, but isn't always. That does imply that something else is getting in the way.

**Reply by 1natsu172** [reply]
Yes, minimum duration hasn't changed almost, sometimes it slows down intensely…Maybe something is getting in the way… 🤷 

If we could see the details in the [clinicjs](https://clinicjs.org/), might be able to get something. (My machine has a problem with [this issue](https://github.com/clinicjs/node-clinic/issues/164) and the clinic is closed lol)

**Reply by 1natsu172** [reply]
Progress report (my own comparison):

After caching `resolveConfig`, build time is stable and blazing fast(before 4s~8s). Browser open time is also 1/2 sometimes, but unstable and something still getting in the way (sometimes it's about 600ms, sometimes about 1.7 s). 

![image](https://github.com/user-attachments/assets/bf33c88d-1cd2-467e-a821-466fef40a64b)

There may be more than one cause. And, after all, there are performance issues with both build time and browser opening time.

**Reply by aklinker1** [reply]
Huh ... `resolveConfig` should only be called once at the start, then again when a file is changed...

Caching it isn't a valid solution, but it is interesting that it speeds up so dramatically when cached.

But yeah, I don't think that's the issue with opening the browser, there's probably something else

**Reply by 1natsu172** [reply]
@aklinker1 I figured out one of the cause! I submitted #919

**Reply by aklinker1** [reply]
Released @1natsu172's PR in [v0.19.6](https://github.com/wxt-dev/wxt/releases/tag/wxt-v0.19.6). Let us know if it fixes your slow down!

**Reply by jaek-s** [reply]
After a quick look, this looks a lot better! After installing it took a little while to open the browser, around 8s. subsequent starts were much quicker with around 700ms to open the browser. Thank you!

**Reply by 1natsu172** [reply]
Yes, I don't know why, but it start up faster the second time around. Something is still lurking. Let's keep 
investigating.

@jaek-s BTW, is your machine MacOS or Windows?

**Reply by jaek-s** [reply]
I'm on MacOS with an M1 chip

**Reply by 1natsu172** [reply]
Thank you. I add it to the sample of information 👍

**Reply by 1natsu172** [reply]
Alright, based on Aaron's bench, I figured out why the performance started to slow down.

https://github.com/wxt-dev/wxt/discussions/866#discussioncomment-10319520
>  <img alt="Screenshot 2024-08-12 at 6 27 30 PM" width="1409" src="https://private-user-images.githubusercontent.com/10101283/357228600-74ea45bb-30cf-4f11-9980-3002a27a49fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM5ODYxODEsIm5iZiI6MTcyMzk4NTg4MSwicGF0aCI6Ii8xMDEwMTI4My8zNTcyMjg2MDAtNzRlYTQ1YmItMzBjZi00ZjExLTk5ODAtMzAwMmEyN2E0OWZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDEyNTgwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2NDFkODA1NTI0ODY2ZGFhNWE1OTdhOWU5MDI3ZjZiZjI1NWYyYTdkZTI4ZDU4OWJkM2IyNjFkNDljNTVmZTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9kGoFkZKmNiaGYR7J9GmmP2cewUFJdxiqPzVF8Sh_CY">
> 
> * [48783ba](https://github.com/wxt-dev/wxt/commit/48783baec32911e49b6177141b97caafc4c4fdb9)
> * [5eb5be7](https://github.com/wxt-dev/wxt/commit/5eb5be7123b6d9074da609d4398baf77fbb1a6cb) (or [f1bb7cb](https://github.com/wxt-dev/wxt/commit/f1bb7cb0c22dc36259c13292b8f11d35e2f3f4cb)?)
> * And maybe [a0971ca](https://github.com/wxt-dev/wxt/commit/a0971cab52a0195b5886390768a08919776b1e8f)?

It was caused by [48783ba](https://github.com/wxt-dev/wxt/commit/48783baec32911e49b6177141b97caafc4c4fdb9) commit that bumped the vite `5.2.8` to `5.3.0`. There was a change https://github.com/vitejs/vite/pull/16453 in `5.2.9` of [CHANGELOG](https://github.com/vitejs/vite/blob/v5.3.0/packages/vite/CHANGELOG.md#529-2024-04-15) that caused this. The change was to make `vite.watcher` not ignore `outDir` when `emptyOutDir` is `false`. So **until `5.2.9`, `outDir` was ignored** and performance was stable.

I have not yet figured out why the slow down accelerated with [5eb5be7](https://github.com/wxt-dev/wxt/commit/5eb5be7123b6d9074da609d4398baf77fbb1a6cb) (or [f1bb7cb](https://github.com/wxt-dev/wxt/commit/f1bb7cb0c22dc36259c13292b8f11d35e2f3f4cb)?) commits. If something is added that increases the output to `outDir` or `.wxt`, If not, there must be a cause other than `vite.watcher`.

**Reply by 1natsu172** [reply]
I looked closer, the commit that started spiking is https://github.com/wxt-dev/wxt/commit/14011c1 . Hmmm……that `vite-node`…

**Comment by 1natsu172** [comment]
On my machine, I was doing some testing and found that even with firefox, occur the second launch faster issue. 

In addition, launching pure chrome with [chrome-launcher](https://github.com/GoogleChrome/chrome-launcher/tree/main) without WXT was almost the same as the wxt@0.19.6 launch speed. And also had the second launch faster issue.

If everyone else is the same, then this second faster problem might be a macOS issue. I also found the interesting thread. (https://www.reddit.com/r/MacOS/comments/1bp7ugd/why_does_it_take_so_long_to_open_apps_for_the/)
